React/Modules/RCTUIManager.m

```diff
@@ -723,6 +727,20 @@ - (void)_amendPendingUIBlocksWithStylePropagationUpdateForShadowView:(RCTShadowV
   }
 }
 
+- (void)addUIManagerObserver:(id<RCTUIManagerObserver>)observer
+{
+  [_uiManagerObserversLock lock];
```

Comment:
I think it's preferable that adding observers is synchronous, with the current impl that adds an observer for each batch this is needed otherwise the observer could be added after the UI operations are completed.